File: Binary_Tree/94.Binary_Tree_Inorder_Traversal/Python/94.Binary_Tree_Inorder_Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        #======================================#
        # Recursion-based DFS traversal method #
        #======================================#

        ############
        #Initialize
        ##### Result array #####
        res_arry = []

        ####################################
        #Recursion-based DFS loop traversal
        self.dfsTraversal(root, res_arry) #Recursion function call (i.e. whole node)

        return res_arry


    def dfsTraversal(self, root, res_arry):
        """
        :type root: TreeNode
        :type res_arry: List[int]
        :rtype: None, void
        """
        #========================================#
        # Recursion-based DFS traversal function #
        #========================================#
        if (not root): #Issue/Boundary-case handling
            return

        ####################
        #Whole process/flow
        self.dfsTraversal((root.left), res_arry) #Recursion function call (i.e. root left-half node)

        res_arry.append((root.val)) #Update/Record

        self.dfsTraversal((root.right), res_arry) #Recursion function call (i.e. root right-half node)

File: Binary_Tree/94.Binary_Tree_Inorder_Traversal/C++/94.Binary_Tree_Inorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void dfsTraversal(TreeNode * root, vector<int> & res_vect)
    {
        //======================================//
        // Input type:                          //
        //   - TreeNode * (root tree-node)      //
        //   - vector<int> ref. (result vector) //
        // Return type:                         //
        //   - void (no return)                 //
        //======================================//
        if (! root) { return; } //Issue, Boundary-case handling

        /*Whole process, flow*/
        dfsTraversal((root->left), res_vect); //Recursion function call (i.e. root left-half node)

        res_vect.push_back((root->val)); //Update, Record

        dfsTraversal((root->right), res_vect); //Recursion function call (i.e. root right-half node)
    }


    vector<int> inorderTraversal(TreeNode * root) 
    {
        //=================================//
        // Input type:                     //
        //   - TreeNode * (root tree-node) //
        // Return type:                    //
        //   - vector<int> (result vector) //
        //=================================//

        /*Initalize*/
        ///// Result vector /////
        vector<int> res_vect;

        /*Recursion-based DFS loop traversal*/
        dfsTraversal(root, res_vect); //Recursion function call (i.e. whole node)

        return res_vect;
    }
};
